src/hardware/pcspeaker_discrete.cpp

```diff
@@ -0,0 +1,491 @@
+	case 0b00:
+		port_b_state.timerGateOutput = false;
+		port_b_state.speakerOutput   = false;
+		AddDelayEntry(newindex, NeutralOr(amp_negative));
+		break;
+	case 0b01:
+		port_b_state.timerGateOutput = false;
+		port_b_state.speakerOutput   = true;
+		AddDelayEntry(newindex, NeutralOr(amp_negative));
+		break;
+	case 0b10:
+		port_b_state.timerGateOutput = true;
+		port_b_state.speakerOutput   = false;
+		AddDelayEntry(newindex, NeutralOr(amp_positive));
+		break;
+	case 0b11:
+		port_b_state.timerGateOutput = true;
+		port_b_state.speakerOutput   = true;
+		AddDelayEntry(newindex, NeutralLastPitOr(amp_positive));
+		break;
+	};
+
+	// Activate the channel after queuing new speaker entries
+	channel->Enable(true);
+}
+
+// After the speaker has idled (as defined when the number of speaker
+// movements is zero) for a short time, wind down the DC-offset, and
+// then halt the channel.
+void PcSpeakerDiscrete::PlayOrFadeout(const uint16_t speaker_movements,
+                                      size_t requested_samples, int16_t *buffer)
+{
+	if (speaker_movements && requested_samples) {
+		static_assert(idle_grace_time_ms > 0, "Algorithm depends on a non-zero grace time");
+		idle_countdown = idle_grace_time_ms;
+		dc_silencer.Reset();
+	} else if (idle_countdown > 0) {
+		idle_countdown--;
+		last_played_sample = buffer[requested_samples - 1];
+	} else if (!dc_silencer.Generate(last_played_sample, requested_samples, buffer)) {
+		channel->Enable(false);
+	}
+
+	channel->AddSamples_m16(check_cast<uint16_t>(requested_samples), buffer);
+}
+
+void PcSpeakerDiscrete::ChannelCallback(uint16_t frames)
+{
+	constexpr uint16_t render_frames = 64;
+	int16_t buf[render_frames];
+
+	ForwardPIT(1);
+	last_index            = 0.0;
+	uint16_t pos          = 0u;
+	auto sample_base      = 0.0;
+	const auto sample_add = (1.0001) / frames;
+
+	auto remaining = frames;
+	while (remaining > 0) {
+		const auto todo = std::min(remaining, render_frames);
+		for (auto i = 0; i < todo; ++i) {
+			auto index = sample_base;
+			sample_base += sample_add;
+			const auto end = sample_base;
+			auto value     = 0.0;
+			while (index < end) {
+				/* Check if there is an upcoming event */
+				if (entries_queued && entries[pos].index <= index) {
+					volwant = entries[pos].vol;
+					pos++;
+					entries_queued--;
+					continue;
+				}
+				double vol_end;
+				if (entries_queued && entries[pos].index < end) {
+					vol_end = entries[pos].index;
+				} else
+					vol_end = end;
+				const auto vol_len = vol_end - index;
+				/* Check if we have to slide the volume */
+				const auto vol_diff = volwant - volcur;
+				if (vol_diff == 0) {
+					value += volcur * vol_len;
+					index += vol_len;
+				} else {
+					// Check how long it will take to goto
+					// new level
+					// TODO: describe the basis for these
+					// magic numbers and their effects
+					constexpr auto spkr_speed = amp_positive *
+					                            2.0 / 0.070;
```

Comment:
What is 0.070? Extract to constant?

Reply:
1 / 10th RMS (square root 2?) Unfortunately.. I don't know.

I originally put that TODO comment there but these are still elusive.

Reply:
This `constexpr` comes from the prior macro:

``` C++
+#define SPKR_SPEED (float)((SPKR_VOLUME*2)/0.070f)
```

which was previously expanded in 11 places.  

Going to leave it as-is.  The TODO comment still stands.